### PR TITLE
Fix vera test verifier error handling (#674)

### LINKED ISSUE #674: vera test --json reports verified/ok for a postcondition failure caught by vera verify --json
Hi! First, thank you for the useful verifier diagnostics here. I may be misunderstanding the intended relationship between `vera test` and `vera verify`, but I found a small case where `vera test --json` reports success/verified for a program that `vera verify --json` rejects and `vera run` fails on at runtime.

Version: `vera 0.0.155`, repo HEAD `08c1197d4d701fc429cc448abb89a8d08b2f7b7c`.

### Minimal repro

```vera
public fn double(@Int -> @Int)
  requires(true)
  ensures(@Int.result == @Int.0 + @Int.0)
  effects(pure)
{
  @Int.0 + @Int.0 + 1
}

public fn main(@Unit -> @Int)
  requires(true)
  ensures(@Int.result == 42)
  effects(pure)
{
  double(21)
}
```

### Observed behavior

`vera check --json wrong_double_initial.vera` succeeds:

```json
{"ok": true, "diagnostics": [], "warnings": []}
```

`vera verify --json wrong_double_initial.vera` exits `1`:

```json
{
  "ok": false,
  "diagnostics": [
    {"error_code": "E500", "description": "Postcondition does not hold in function 'double'."}
  ],
  "verification": {"tier1_verified": 3, "tier3_runtime": 0, "total": 3}
}
```

This looks expected: the body returns `x + x + 1`, while the postcondition claims `x + x`.

`vera run wrong_double_initial.vera` exits `1`:

```text
Error: Postcondition violation in double(@Int -> @Int)
  ensures(@Int.result == @Int.0 + @Int.0) failed
```

`vera test --json wrong_double_initial.vera` exits `0`:

```json
{
  "ok": true,
  "functions": [
    {"name": "double", "category": "verified", "reason": "Tier 1 (proved)", "trials_run": 0},
    {"name": "main", "category": "verified", "reason": "Tier 1 (proved)", "trials_run": 0}
  ],
  "summary": {"verified": 2, "tested": 0, "total_trials": 0, "total_failures": 0}
}
```

### Expected behavior

I would expect `vera test --json` not to return a green `ok: true` / `verified` result for a program whose contract verification fails with `vera verify --json`.

Possible ways to avoid the false-green signal:

- `test --json` invokes/reuses the same verification gate and returns nonzero with the `E500` diagnostic; or
- `test --json` reports these functions as unverified/failed/skipped rather than `category: "verified"`; or
- if `test` intentionally has narrower semantics than `verify`, the JSON output makes that boundary explicit so downstream automation does not treat this as proof success.

### Why this matters

JSON status is likely to be consumed mechanically by CI pipelines and downstream tooling. A false-green `ok: true` from `test --json`, especially with `category: "verified"` and `reason: "Tier 1 (proved)"`, can cause a build step to accept a contract-incorrect program even though `verify --json` correctly rejects it.

### Question

I am treating `vera verify --json` as the authoritative contract/proof gate. If `vera test` is intentionally meant to skip or only partially reuse verification in this situation, the bug may be in my expectation or in the JSON wording (`verified` / `Tier 1 (proved)`) rather than in the underlying test behavior.

Would you expect `vera test --json` to be a safe green gate for contract correctness, or should downstream tools always run `vera verify --json` separately and treat `test --json` as corroborating/runtime evidence only?

### PR DESCRIPTION
**Pull request by rzyns**
## Summary

Fixes #674.

`vera test` now fails closed when the verifier reports E500/E501/E502 diagnostics instead of letting verifier-refuted functions fall through as `"verified" / "Tier 1 (proved)"`.

This makes `vera test --json` safe as a green contract gate again for the verifier-error cases discussed in #674.

## Behavior changes

- Preserve verifier error diagnostics in `TestResult.diagnostics`.
- Add/handle a `"failed"` function category for verifier-refuted public functions.
- Surface `TestSummary.failed` in JSON and human output.
- Return non-zero from `vera test` / `vera test --json` when result diagnostics contain errors.
- Set JSON `"ok": false` when verifier errors are present.
- Attribute:
  - E500 postcondition failures to the function named in `in function '...'`
  - E501 call-site precondition failures to the caller, not the callee
  - E502 `@Nat` subtraction underflow to the function named in `in '...'`
- Keep `--fn` rows filtered to the selected function while still failing closed on whole-file verifier errors.
- Surface private/non-displayed verifier failures in diagnostics and human summaries instead of silently dropping them.
- Keep Tier 3 E700 runtime contract failures distinct from static verifier errors in human summaries.
- Avoid double-counting multiple verifier diagnostics that are already represented by a displayed failed function.

## Tests added / updated

Regression coverage now includes:

- `vera test --json` returns nonzero and `ok: false` for E500 verifier errors.
- Human output shows verifier-refuted functions as `FAILED`.
- Private verifier failures are reported even when no private function row is displayed.
- `--fn` keeps function rows filtered but still fails closed on unselected whole-file verifier errors.
- E501 is attributed to the caller rather than the callee.
- E502 `@Nat` underflow is classified as a failed verifier error.
- Static verifier failures are summarized separately from Tier 3 runtime trial failures.
- Tier 3 E700 runtime failures are not summarized as verifier errors.
- Multiple verifier diagnostics on one displayed failed function are not double-counted as unlisted verifier errors.
- CLI dispatch for `vera test --json` returns nonzero for verifier errors.

## Verification

Ran locally:

```text
pytest tests/test_tester.py tests/test_tester_coverage.py tests/test_cli.py -q
276 passed
```

```text
pytest tests/ -q
3886 passed, 14 skipped, 16 deselected
```

```text
mypy vera/
Success: no issues found in 59 source files
```

```text
python scripts/check_conformance.py
All 86 conformance programs pass.
```

```text
python scripts/check_examples.py
All 34 examples pass (check + verify).
```

```text
pytest tests/test_conformance.py -q
416 passed, 14 skipped
```

Also checked:

```text
git diff --check
```

and a staged added-line secret scan; both were clean.

## Notes

This follows the intended direction from the maintainer comment on #674:

- harvest E500/E501/E502 verifier errors,
- introduce `"failed"`,
- make failed status take precedence over verified/Tier 3 classification,
- surface `TestSummary.failed`,
- return non-zero,
- and add regression coverage.

I kept the displayed `--fn` function list scoped to the selected function, but made diagnostics/exit status reflect whole-file verifier errors so `--fn` cannot false-green a file with verifier-refuted contracts.

## Summary by CodeRabbit

* **Bug Fixes**
  * Verifier failures (E500, E501, E502) now cause tests to fail with non‑zero exit codes, JSON ok:false and preserved diagnostics, and human output shows failed functions plus a Diagnostics section; --fn filtering still surfaces whole‑file verifier errors.

* **New Features**
  * Improved error attribution and separate counts for static verifier failures vs runtime (Tier‑3) failures; avoids double‑counting multiple diagnostics; E700 remains distinct.

* **Tests**
  * Expanded regression and integration coverage for JSON/human output, filtering, attribution and exit codes.

* **Documentation**
  * CHANGELOG, KNOWN_ISSUES, README, ROADMAP and TESTING updated.

[Review Change Stack](https://app.coderabbit.ai/change-stack/aallan/vera/pull/685?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: tests/test_tester.py [hunk 1]
```diff
@@ -41,6 +41,78 @@ def _fn_result(result: TestResult, name: str) -> FunctionTestResult:
...
+  effects(pure)
+{
+  needs_pos(@Int.0)
+}
+"""
+        result = _test(source)
+        needs_pos = _fn_result(result, "needs_pos")
+        caller = _fn_result(result, "caller")
+        assert needs_pos.category != "failed"
+        assert caller.category == "failed"
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_tester.py:97]: @coderabbitai fixed in 8d12298 — strengthened to `assert needs_pos.category == "verified"`.

You're right that the negative `!= "failed"` would still pass if `needs_pos` were classified as `"skipped"` or any other non-failed bucket, which wouldn't actually pin the E501-attribution-to-caller invariant. The positive form proves the callee is correctly proven AND that the regression is contained to the caller.

Added a comment at the assertion site explaining the rationale so future contributors don't weaken it back to a negative check.

File: vera/tester.py [hunk 2]
```diff
@@ -398,15 +414,21 @@ def _classify_functions(
...
     for diag in verify_diagnostics:
         if diag.severity == "warning" and diag.error_code in tier3_codes:
             # Extract fn name from description: '...'
             m = re.search(r"'(\w+)'", diag.description)
             if m:
                 tier3_fns.add(m.group(1))
+        elif diag.severity == "error" and diag.error_code in verification_error_codes:
+            name = _failed_function_name(diag)
+            if name:
+                failed_fns[name] = diag.error_code or "verification error"
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/tester.py]: @coderabbitai fixed in 8d12298 — switched to first-hit semantics (`if name not in failed_fns`).

The order-dependence you identified is real and exercised by the existing `test_displayed_failed_function_does_not_double_count_verifier_errors` test, which uses a function attracting both E500 (`ensures(false)` is unsatisfiable) and E502 (`@Nat - @Nat` underflow). Pre-fix the displayed `reason` would flip with diagnostic iteration order; post-fix it's pinned to whichever diagnostic the verifier emitted first.

Went with first-hit over an explicit priority list because:
1. For these three codes there's no meaningful "should be primary" ranking — each names a distinct contract violation
2. `setdefault`-style semantics are minimum-surprise: "first one wins, document the rule"
3. A future priority list could be added if a real use case for it emerges; YAGNI for now

Added an inline comment explaining the rule so a future contributor doesn't accidentally revert it (e.g. by reformatting to a one-liner that loses the `not in` check).

Two related nits also fixed in the same commit since they touch the same code:
- Extracted `VERIFICATION_ERROR_CODES = frozenset({"E500", "E501", "E502"})` as a module-level constant in `vera/tester.py` — was duplicated between `tester.py` and `cli.py`
- Renamed `_failed_function_name` → `failed_function_name` (drop underscore) — was being imported from `cli.py`, so the private-by-convention prefix was misleading

Plus a couple of housekeeping items that came up alongside: doc-counts gate (CI was failing on 7 stale numbers in TESTING.md/ROADMAP.md/README.md after the new tests landed) and the empty `| Bug | Issue |` table in `KNOWN_ISSUES.md` (project convention is to keep the heading and write "No known bugs." in the body).

File: tests/test_cli.py [hunk 3]
```diff
@@ -1376,6 +1376,200 @@ def test_json_output(self, capsys: pytest.CaptureFixture[str]) -> None:
...
+}
+
+public fn main(@Unit -> @Int)
+  requires(true)
+  ensures(@Int.result == 42)
+  effects(pure)
+{
+  double(21)
+}
+"""))
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_cli.py:1406]: @coderabbitai fixed in 8d12298 — added `encoding="utf-8"` to `_bad_vera`. The fix benefits every caller of the helper (30+ sites in `test_cli.py`), not just rzyns's new tests. CLAUDE.md "Cross-platform pitfalls" already documents this as a required convention for test fixtures, so it's the right place to put the encoding rather than at each call site.

**Review by aallan — DISMISSED** [review]
Thanks @rzyns — this is a fine fix. Approving once CI clears.

A few notes on what's here, for anyone looking at the PR:

**Correctness.** The classifier change in `vera/tester.py` is right: verifier-refuted contracts now take precedence over the verified/tier3/skipped buckets, so a function the verifier can refute statically can't fall through to a green-test result. The E501 attribution to the caller (not the callee) is the right call — the missing-precondition obligation lives at the caller, and that's where the user needs to add the `requires(...)` to make the verifier accept the call. The `--fn` filter logic is also nicely handled: filtered output stays scoped to the selected function, but whole-file verifier errors still fail closed (otherwise `vera test --fn safe_fn buggy_file.vera` would false-green).

**Tests.** Eleven new regression tests across `test_tester.py` and `test_cli.py`, each pinning a distinct property — JSON output, human output, private function visibility, `--fn` filter behaviour, summary split, Tier-3-vs-static distinction, double-count avoidance, dispatcher contract. The double-count test (function attracting both E500 and E502 on the same body via `ensures(false)` + `@Nat - @Nat`) was an edge case I'd have missed.

**Cleanup commit (8d12298).** I pushed a follow-up handling the three CR comments that arrived later — `_bad_vera` encoding (now `encoding="utf-8"` — benefits 30+ callers, not just the new tests), the `needs_pos.category != "failed"` → `== "verified"` assertion strengthening (the negative form would still pass under `"skipped"`), and the `failed_fns` dict-clobber fix (first-hit semantics, so multi-diagnostic functions get a deterministic `reason` instead of one that flips with iteration order). Also rolled in a couple of nits: extracted `VERIFICATION_ERROR_CODES` as a module-level `frozenset` constant (was duplicated between `tester.py` and `cli.py`), and dropped the underscore from `failed_function_name` since `cli.py` imports it across modules. Plus the doc-counts CI gate (7 stale numbers in TESTING.md / ROADMAP.md / README.md after the test count went 3,905 → 3,916) and the `| Bug | Issue |` empty-table convention in `KNOWN_ISSUES.md`.


Thanks for the well-shaped contributions across #674, #675, and this one.

File: tests/test_cli.py [hunk 4]
```diff
@@ -1376,6 +1383,200 @@ def test_json_output(self, capsys: pytest.CaptureFixture[str]) -> None:
...
+        rc = cmd_test(str(path), as_json=True, fn_name="identity")
+        assert rc == 1
+        data = json.loads(capsys.readouterr().out)
+        assert data["ok"] is False
+        assert data["summary"]["failed"] == 0
+        assert any(d["error_code"] == "E500" for d in data["diagnostics"])
+        assert [f["name"] for f in data["functions"]] == ["identity"]
+        assert data["functions"][0]["category"] == "verified"
+
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_cli.py:1514]: @coderabbitai fixed in a7b0c61 — added `test_fn_filter_surfaces_unselected_verifier_error_text` as a sibling of the JSON test. Pins the same fail-closed invariant against human output: displayed function list stays scoped to `identity` via `--fn`, `broken`'s verifier error still surfaces in the Diagnostics section, summary line names the unlisted verifier error, and exit code is 1.

Also strengthened the existing JSON test in the same commit to assert `data["summary"]["unlisted_errors"] == 1` — pins the new structured field added per the other finding in this round.

File: vera/tester.py [hunk 5]
```diff
@@ -87,6 +87,12 @@ class TestResult:
 # Types that require the raw_args calling convention (String uses i32_pair ABI)
 _NEEDS_RAW = {STRING, FLOAT64}
 
+# Verifier-error codes that cause a function to be classified as
+# ``"failed"`` rather than ``"verified"`` / ``"tier3"`` / ``"skipped"``.
+# Exported (no underscore) because ``vera/cli.py`` also uses this set
+# in its summary-line de-duplication logic.
+VERIFICATION_ERROR_CODES = frozenset({"E500", "E501", "E502"})
+
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/tester.py:105]: @coderabbitai went with option (c) in a7b0c61 — expose structured data on `TestSummary` rather than widening the public API.

You're right that the round-1 rename was the wrong shape. `VERIFICATION_ERROR_CODES` and `failed_function_name` were CLI-only bookkeeping that didn't belong on `vera.tester`'s public surface; option (a) (just drop the underscore) left implementation details leaking across the module boundary. Option (b) (a shared helper) would have addressed the symptom but kept the underlying coupling. Option (c) makes the engine the source of truth and `cli.py` purely presentational.

Specifically:
- `TestSummary` gains `unlisted_errors: int = 0` — count of verifier-error diagnostics whose attributable function isn't in the displayed `functions` list
- `_TestEngine.run()` computes it after the main loop by comparing the diagnostic-attribution set against `displayed_failed` (functions with `category == "failed"`)
- `VERIFICATION_ERROR_CODES` → `_VERIFICATION_ERROR_CODES` (private)
- `failed_function_name` → `_failed_function_name` (private)
- `vera/cli.py` drops both imports + the local `displayed_failed`/`unlisted_errors` computation; reads `s.unlisted_errors` directly
- JSON output gains `"unlisted_errors": s.unlisted_errors` in the summary dict so downstream CI consumers can read the structured count (instead of re-running attribution on the diagnostics array)
- Strengthened `test_fn_filter_surfaces_unselected_verifier_error` (JSON) to assert the new field's value (= 1 in that scenario)

Net effect: `vera.tester` has a smaller public API, `vera/cli.py` has no classifier logic (just formatting), and downstream JSON consumers get the answer as structured data instead of having to derive it. Wins on all three axes.

**Comment by aallan** [comment]
**Cleanup commit.** (`a7b0c61`) — round-2 CR findings: text-mode sibling for the `--fn` filter test, and restructuring the public API. CR (correctly) pushed back on the round-1 rename: those classifier internals were CLI-only bookkeeping that didn't belong on `vera.tester`'s public surface. Replaced with `TestSummary.unlisted_errors` as structured data, the engine computes the value, `cli.py` reads it. `_VERIFICATION_ERROR_CODES` and `_failed_function_name` are private again. JSON output gains the new field so downstream CI consumers read structured data instead of re-running attribution on the diagnostics array.
